common, osd: kill OpQueue::length() to minimize inter-core traffic.

rzarzynski opened the pull request:
In the case of `WeightedPriorityQueue` the `size` manipulation was a subject to ping-pong between a `tp_osd_tp` and messenger threads.

Be aware that the profiling has been during the work slabbed on containers. In current `master` the stall might be just overlapped. Thus, I'm sending this patch rather as a clean-up. Killing `size` in other `OpQueue`'s implementations is TODO.

Before
------
```
-    1,50%     1,34%  msgr-worker-0  ceph-osd  [.] WeightedPriorityQueue<OpQueueItem, unsigned long>::Queue::insert
```

Where self if 1.34%
```
       │       mov    %r12d,0x2c(%r14)
       │               }
       │               ++size;
 17,33 │1ba:   addl   $0x1,0x2ec48(%r14)
       │             }
```

After
-----
```
-    1,05%     0,93%  msgr-worker-0  ceph-osd  [.] WeightedPriorityQueue<OpQueueItem, unsigned long>::Queue::insert
```

Let's annotate hottest place (self is 0,93%):
```
       │           //Find the upper bound, cache the previous value and if we should
       │           //store it in the left or right node
       │           bool left_child = true;
       │           while(x){
       │ a0:   mov    %rax,%rbx
       │              ++depth;
       │              y = x;
       │              x = (left_child = comp(key, x)) ?
  8,86 │ a3:   cmp    0x20(%rbx),%r15
       │     _ZN5boost9intrusive31default_rbtree_node_traits_implIPvE9get_rightERKPNS0_11r
```

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

liewegas commented on src/common/mClockPriorityQueue.h:
```diff
@@ -68,7 +68,7 @@ namespace ceph {
 
       unsigned tokens, max_tokens;
-      int64_t size;
+      int64_t size; // XXX: this is only for the sake of dump().
```
kill this field then?

aclamk commented on src/common/mClockPriorityQueue.h:
```diff
@@ -229,7 +229,8 @@ namespace ceph {
-    unsigned length() const override final {
+    // XXX: used only by the unitest?
+    unsigned length() const {
       unsigned total = 0;
```
This is true, only unittests use length().

liewegas commented:
needs rebase

rzarzynski commented:
Rebased to `master` and introduced `::get_size_slow()` for unit  tests. Additionally pushed the follow-up commit which drops the `size` field from `mClockOpClassQueue`, `mClockClientQueue` and `mClockQueue` as well.
